### PR TITLE
Lower multi-dimensional IRFFT on GPU to a single C2R transform again

### LINKED ISSUE #40143: GPU: multi-dimensional `irfft` lowering (since 0.10.0) is ~1.7× slower than the single cuFFT C2R op for Hermitian-symmetric inputs — request an opt-in for the direct path
Since jax 0.10.0 (by #36383 on #29325), `jax/_src/lax/fft.py::_fft_lowering_gpu` (registered for `cuda` and
`rocm`) decomposes every multi-dimensional `IRFFT` into

```
transpose → complex IFFT over the outer axes → transpose → 1-D IRFFT on the last axis
```

so that GPU results match NumPy, which requires Hermitian symmetry only along the last
axis (cuFFT's multi-dimensional C2R assumes it on every transformed axis). That is the
right default. But for inputs that *are* Hermitian-symmetric on every axis — the spectrum
of a real array times a real centro-symmetric filter, i.e. the inner loop of every
FFT-convolution renderer — the two lowerings give identical results and the decomposition
is pure cost: two cuFFT launches and two transposes instead of one multi-dimensional C2R.
We measured it as a 1.7× slowdown of the transform and a 25–31% slowdown of a
real workload (a forced-photometry template renderer) between jax 0.9.0 and 0.10.0, on
both an H100 and an L40S, reproduced bit-for-bit equal outputs on both paths, and
bisected it to exactly this rule.

**Reproduction** ([jax_irfft_gpu_repro.py](https://github.com/user-attachments/files/31342376/jax_irfft_gpu_repro.py); jax + numpy only)

The script builds a Hermitian-symmetric input (`rfft2` of random real images × a
Gaussian PSF spectrum, layout `(49, 262, 80, 41)` complex64), times the default
`jnp.fft.irfft2(x, s=(80, 80))`, prints the HLO `fft` ops, then — for comparison only —
re-registers the generic lowering for the GPU platform to time the single-op C2R path on
the same bytes and checks the outputs.

```
$ CUDA_VISIBLE_DEVICES=1 python jax_irfft_gpu_repro.py          # L40S, card shared with another job
jax 0.11.1 | gpu NVIDIA L40S
input: (49, 262, 80, 41) complex64 (rfft2 layout of 49x262 real 80x80 images)
default lowering : HLO fft ops [('IFFT', '80'), ('IRFFT', '80')], transposes 2,    9.88 ms
direct C2R (opt) : HLO fft ops [('IRFFT', '80,80')], transposes 0,    4.59 ms
ratio default/direct = 2.16x ; max|diff|/max|y| = 2.3e-07

$ ... same script under jax 0.9.0
jax 0.9.0 | gpu NVIDIA L40S
default lowering : HLO fft ops [('IRFFT', '80,80')], transposes 0,    5.02 ms
no _fft_lowering_gpu in this jax (<0.10): default IS the single-op path
```

(The 2.3e-7 relative difference is float32 rounding between the two cuFFT
algorithms; on our production spectra — products of `rfft2` transforms — the two
paths were bit-identical.)

**Measurements** (same captured inputs, jitted, min of 7, `block_until_ready`; H100
80 GB, allocation-exclusive; jax wheels `jax[cuda12]==V`, python 3.12):

| jax | `irfft2`-heavy render, 49×262 stamps 80² | 49×16 grids 160² | HLO for one `irfft2` |
| --- | --- | --- | --- |
| 0.9.0 | 1.43 ms | 3.85 ms | 1 op: `fft_type=IRFFT, fft_length={80,80}` |
| 0.10.0 | 1.82 ms (+27%) | 5.05 ms (+31%) | `IFFT {80}` + transpose + `IRFFT {80}` (+2 transposes) |
| 0.11.1 | same as 0.10.0 |     | same |

(0.6.0–0.8.0 behave like 0.9.0; 0.5.3 within 5% of it.) Isolated transform on an L40S,
jax 0.11.1 vs 0.9.0: `(49,262,80,41)→(80,80)` 8.2 ms vs 4.8 ms (1.7×); `(49,16,160,81)→
(160,160)` 3.8 vs 0.9 ms; `max|Δ|/max|y| = 0` between the two lowerings on these inputs.
`jax.lax.fft(x, FftType.IRFFT, (80, 80))` lowers through the same rule, so there is no
user-level way to request the direct C2R.

**Proposal**

1. An opt-in for callers who guarantee full Hermitian symmetry, e.g. a config flag
  `jax_gpu_irfft_assume_hermitian` (default `False`, NumPy semantics unchanged) checked at
  the top of `_fft_lowering_gpu`, or an argument threaded through `lax.fft` /
  `jnp.fft.irfftn` — one branch plus tests (equality on symmetric inputs under both
  paths; NumPy agreement on a non-symmetric input with the flag off). Happy to send the PR.
2. Longer term, an exact lowering without the transposes (outer-axis C2C via strided
  cuFFT plans on non-trailing dims, which needs XLA's FFT thunk to accept them), or a
  single full C2C IFFT of the Hermitian-extended last axis plus the real part (one launch,
  no transposes, ~2× the FFT traffic) — worth measuring.

**Environment**: NVIDIA H100 80GB HBM3 (CUDA 12.9 wheels) and L40S (CUDA 12.8 wheels),
Linux x86_64, python 3.12/3.13, `jax[cuda12]` 0.5.3 … 0.11.1 from PyPI.

### PR DESCRIPTION
**Pull request by hbahk**
Fixes #40143. Related: #29325.

Since 0533a2068 (0.10.0) the GPU lowering decomposes a multi-dimensional IRFFT into an IFFT over the outer axes and a 1-D IRFFT, with two transposes, so that results match NumPy/CPU/TPU (Hermitian symmetry assumed only along the last axis) rather than cuFFT's multi-dimensional C2R (assumed along every axis). The decomposition costs ~1.7× the time of one multi-dimensional C2R.

This PR keeps the NumPy semantics and restores the single C2R: the input is first made Hermitian-symmetric along the outer axes, which does not change its IRFFT.

*Why it is exact.* An array in RFFT layout is the RFFT of some real array iff its slices at index 0 and, for an even last length n, at n//2 along the last axis are Hermitian-symmetric along the outer FFT axes; the other entries are unconstrained. NumPy's `irfftn` (IFFT over the outer axes, then a 1-D IRFFT that reads only the real part of the DC and Nyquist bins) is `(1/N) Σ_{k∈half} m_k Re(z_k e^{+2πi k·j/N})`, which depends on those two slices only through their Hermitian-symmetric part. Replacing each slice by `(z_k + conj(z_{-k}))/2` therefore leaves the result unchanged under NumPy's convention, and yields a valid half-spectrum on which a C2R transform that assumes symmetry on every axis computes the same inverse. The symmetrization is a single `concatenate` (one elementwise fusion over the input, which fuses with a producing elementwise op); updating the slices in place makes XLA copy the operand and was slower.

*Measurements* (L40S, cuFFT 11.4, jaxlib 0.11.1, complex64, min of 15; input as a jit parameter / as an intermediate of the same jit):

| irfftn | before (0.10.0–0.11.1) | this PR | bare C2R, no symmetrization |
|---|---:|---:|---:|
| `(49, 262, 80, 41) → (80, 80)` | 7.29 / 7.29 ms | 5.25 / 5.24 ms | 4.17 / 5.18 ms |
| `(49, 16, 160, 81) → (160, 160)` | 1.56 / 1.56 ms | 1.09 / 1.09 ms | 0.85 / 1.08 ms |
| `(8, 32, 32, 17) → (32, 32, 32)` | 0.05 ms | 0.04–0.05 ms | 0.05 ms |

`max|new − old| / max|old| = 2e-7` (rounding). Outputs agree with NumPy to float32 rounding for inputs that are not symmetric along the outer axes, including the `s=(5, 1)` case of #29325 (where the bare cuFFT C2R does differ), and gradients (GPU == CPU) are unchanged.

*Tests.* `testIrfftnNonHermitianOuterAxes` (12 shape/`s` combinations incl. last lengths 1 and 2, 3-D, batch dims, against NumPy) and `testIrfftnGpuLoweringIsASingleFft` (lowers for `cuda` via `lowering_platforms`, so it runs on CPU CI: one `stablehlo.fft … IRFFT, length = [6, 8]`, no transpose); the existing `testIrfftnNonSymmetricOuterAxes` is kept. `tests/fft_test.py` passes on CPU and on an L40S with and without `JAX_ENABLE_X64`. The lowering is the same for `rocm` (not tested here). CHANGELOG entry added.

### CODE REVIEW

**Review by hawkinsp — APPROVED** [review]
Thanks for the PR!
